src/shim/shim.c

```diff
@@ -145,6 +144,22 @@ _shim_load() {
     shim_enableInterposition();
 }
 
+// This function should be called before any wrapped syscall. We also use the
+// constructor attribute to be completely sure that it's called before main.
+__attribute__((constructor)) void shim_ensure_init() {
+    static __thread bool in_init = false;
```

Comment:
We can't use `thread_local` here, can we? I think that's more portable if all of our compilers support it.